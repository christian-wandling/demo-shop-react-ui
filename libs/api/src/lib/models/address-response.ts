/* tslint:disable */
/* eslint-disable */
/**
 * Demo Shop API
 * A comprehensive API for managing an online store, providing endpoints for product catalog, user management, shopping cart operations, and order processing
 *
 * The version of the OpenAPI document: v1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { mapValues } from '../runtime';
/**
 *
 * @export
 * @interface AddressResponse
 */
export interface AddressResponse {
  /**
   *
   * @type {string}
   * @memberof AddressResponse
   */
  street: string;
  /**
   *
   * @type {string}
   * @memberof AddressResponse
   */
  apartment: string;
  /**
   *
   * @type {string}
   * @memberof AddressResponse
   */
  city: string;
  /**
   *
   * @type {string}
   * @memberof AddressResponse
   */
  region?: string | null;
  /**
   *
   * @type {string}
   * @memberof AddressResponse
   */
  zip: string;
  /**
   *
   * @type {string}
   * @memberof AddressResponse
   */
  country: string;
}

/**
 * Check if a given object implements the AddressResponse interface.
 */
export function instanceOfAddressResponse(value: object): value is AddressResponse {
  if (!('street' in value) || value['street'] === undefined) return false;
  if (!('apartment' in value) || value['apartment'] === undefined) return false;
  if (!('city' in value) || value['city'] === undefined) return false;
  if (!('zip' in value) || value['zip'] === undefined) return false;
  if (!('country' in value) || value['country'] === undefined) return false;
  return true;
}

export function AddressResponseFromJSON(json: any): AddressResponse {
  return AddressResponseFromJSONTyped(json, false);
}

export function AddressResponseFromJSONTyped(json: any, ignoreDiscriminator: boolean): AddressResponse {
  if (json == null) {
    return json;
  }
  return {
    street: json['street'],
    apartment: json['apartment'],
    city: json['city'],
    region: json['region'] == null ? undefined : json['region'],
    zip: json['zip'],
    country: json['country'],
  };
}

export function AddressResponseToJSON(json: any): AddressResponse {
  return AddressResponseToJSONTyped(json, false);
}

export function AddressResponseToJSONTyped(value?: AddressResponse | null, ignoreDiscriminator: boolean = false): any {
  if (value == null) {
    return value;
  }

  return {
    street: value['street'],
    apartment: value['apartment'],
    city: value['city'],
    region: value['region'],
    zip: value['zip'],
    country: value['country'],
  };
}
