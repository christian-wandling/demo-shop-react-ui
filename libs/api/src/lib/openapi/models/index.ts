/* tslint:disable */
 
export * from './add-cart-item-request';
export * from './address-response';
export * from './cart-item-response';
export * from './image-response';
export * from './order-item-response';
export * from './order-list-response';
export * from './order-response';
export * from './order-status';
export * from './product-list-response';
export * from './product-response';
export * from './shopping-session-response';
export * from './update-cart-item-quantity-request';
export * from './update-user-address-request';
export * from './update-user-phone-request';
export * from './user-response';
