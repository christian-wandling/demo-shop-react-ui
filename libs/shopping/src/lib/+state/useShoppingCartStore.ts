import { create } from 'zustand';
import { AddCartItemRequest, CartItemResponse, getApi, UpdateCartItemQuantityRequest } from '@demo-shop-react-ui/api';

interface ShoppingCartState {
  items: CartItemResponse[];
  sessionId: number | null;
  showCart: boolean;
  fetchCurrentSession: () => Promise<void>;
  add: (addCartItemRequest: AddCartItemRequest) => Promise<void>;
  remove: (id: number) => Promise<void>;
  updateItemQuantity: (id: number, updateCartItemQuantityRequest: UpdateCartItemQuantityRequest) => Promise<void>;
  getItemById: (id: number) => CartItemResponse | undefined;
  getItemByProductId: (productId: number) => CartItemResponse | undefined;
  setShowCart: (showCart: boolean) => void;

  getTotalPrice: () => number;

  getItemCount: () => number;

  hasActiveSession: () => boolean;
}

export const useShoppingCartStore = create<ShoppingCartState>()((set, get) => {
  return {
    items: [],
    sessionId: null,
    showCart: false,

    getTotalPrice(): number {
      return get().items.reduce((sum, item) => sum + item.totalPrice, 0);
    },

    getItemCount(): number {
      return get().items.reduce((sum, item) => sum + item.quantity, 0);
    },

    hasActiveSession(): boolean {
      return get().sessionId !== null;
    },

    fetchCurrentSession: async () => {
      const { shoppingSessionApi } = getApi();
      const session = await shoppingSessionApi.resolveCurrentShoppingSession();
      set({ sessionId: session.id, items: session.items });
    },

    add: async (addCartItemRequest: AddCartItemRequest) => {
      if (!get().hasActiveSession()) {
        throw new Error('No active shopping session');
      }

      const itemInCart = get().getItemByProductId(addCartItemRequest.productId);

      if (itemInCart) {
        get().updateItemQuantity(itemInCart.id, { quantity: itemInCart.quantity + 1 });
        return;
      }

      const { shoppingSessionApi } = getApi();
      await shoppingSessionApi.addCartItem({ addCartItemRequest });
      get().fetchCurrentSession();
    },

    remove: async (id: number) => {
      if (!get().hasActiveSession()) {
        throw new Error('No active shopping session');
      }

      const itemInCart = get().getItemById(id);

      if (!itemInCart) {
        throw new Error('Item not found');
      }

      if (itemInCart.quantity > 1) {
        get().updateItemQuantity(itemInCart.id, { quantity: itemInCart.quantity - 1 });
        return;
      }

      const { shoppingSessionApi } = getApi();
      await shoppingSessionApi.removeCartItem({ id });
      get().fetchCurrentSession();
    },

    updateItemQuantity: async (id: number, updateCartItemQuantityRequest: UpdateCartItemQuantityRequest) => {
      if (!get().hasActiveSession()) {
        throw new Error('No active shopping session');
      }

      const { shoppingSessionApi } = getApi();
      await shoppingSessionApi.updateCartItemQuantity({ id, updateCartItemQuantityRequest });
      get().fetchCurrentSession();
    },

    getItemById: id => {
      return get().items.find(item => item.id === id);
    },

    getItemByProductId: productId => {
      return get().items.find(item => item.productId === productId);
    },

    setShowCart: (showCart: boolean) => {
      set(state => ({ ...state, showCart }));
    },
  };
});
