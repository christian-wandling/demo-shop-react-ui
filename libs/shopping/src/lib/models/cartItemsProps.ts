import { CartItemResponse } from '@demo-shop-react-ui/api';

export interface CartItemsProps {
  /**
   * The list of items in the cart to be displayed.
   */
  items: CartItemResponse[];

  /**
   * Callback fired when a user requests to remove an item from the cart.
   * The callback receives the ID of the item to be removed.
   */
  onRemoveItem: (itemId: number) => void;

  onThumbnailClick?: () => void;
}
