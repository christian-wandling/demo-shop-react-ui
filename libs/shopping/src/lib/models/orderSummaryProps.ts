import { CartItemResponse } from '@demo-shop-react-ui/api';

export interface OrderSummaryProps {
  items: Array<CartItemResponse>;
  totalPrice: number;
  checkoutButtonEnabled: boolean;
  onRemoveItem: (id: number) => void;
  onCheckout: () => void;
}
