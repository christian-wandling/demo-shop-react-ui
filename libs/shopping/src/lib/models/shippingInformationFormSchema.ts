import { z } from 'zod';

export const shippingInformationFormSchema = z.object({
  firstname: z.string().min(1, 'First name is required').readonly(),
  lastname: z.string().min(1, 'Last name is required').readonly(),
  email: z.string().email('Invalid email address').readonly(),
  phone: z.string().nullable(),
  address: z
    .object({
      street: z.string().min(1, 'Street is required'),
      apartment: z.string().min(1, 'Apartment is required'),
      city: z.string().min(1, 'City is required'),
      region: z.string(),
      zip: z.string().min(1, 'Zip is required'),
      country: z.string().min(1, 'Country is required'),
    })
    .required(),
});

export type ShippingInformationFormData = z.infer<typeof shippingInformationFormSchema>;
