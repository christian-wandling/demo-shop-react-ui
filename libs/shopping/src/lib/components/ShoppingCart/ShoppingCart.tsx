import { useMemo } from 'react';
import { Link } from 'react-router';
import { AnimatePresence, motion } from 'framer-motion';
import { CartItems } from '../CartItems/CartItems';
import { useShoppingCartStore } from '../../+state/useShoppingCartStore';
import { formatCurrency } from '@demo-shop-react-ui/shared';

/**
 * Component that displays the shopping cart with animation effects.
 * Shows cart items, total price, and provides checkout functionality.
 *
 * @example
 * <ShoppingCart />
 */
export const ShoppingCart = () => {
  const items = useShoppingCartStore(state => state.items);
  const showCart = useShoppingCartStore(state => state.showCart);
  const getTotalPrice = useShoppingCartStore(state => state.getTotalPrice);
  const remove = useShoppingCartStore(state => state.remove);
  const setShowCart = useShoppingCartStore(state => state.setShowCart);

  /**
   * Determines if the checkout button should be enabled
   * Button is enabled only when there are items in the cart
   */
  const checkoutButtonEnabled = useMemo(() => items.length > 0, [items]);

  /**
   * Closes the cart slide-over panel
   * Updates the showCart state to hide the cart UI
   */
  const closeCart = () => {
    setShowCart(false);
  };

  if (!showCart) return null;

  return (
    <AnimatePresence>
      <motion.div
        key="backdrop"
        initial={{ opacity: 0 }}
        animate={{ opacity: 1 }}
        exit={{ opacity: 0 }}
        transition={{ duration: 0.5 }}
        onClick={closeCart}
        onKeyDown={e => e.key === 'Escape' && closeCart()}
        className="relative z-50"
        role="dialog"
        aria-labelledby="slide-over-title"
        aria-modal="true">
        <div className="fixed inset-0 bg-gray-500 bg-opacity-75 transition-opacity" aria-hidden="true"></div>
        <div className="fixed inset-0 overflow-hidden">
          <div className="absolute inset-0 overflow-hidden"></div>
        </div>
      </motion.div>

      <motion.div
        key="slideover"
        initial={{ transform: 'translateX(100%)' }}
        animate={{ transform: 'translateX(0)' }}
        exit={{ transform: 'translateX(100%)' }}
        transition={{ duration: 0.7 }}
        className="z-50 pointer-events-none fixed inset-y-0 right-0 flex max-w-full pl-10">
        <div className="pointer-events-auto w-screen max-w-md">
          <div className="flex h-full flex-col bg-white shadow-xl">
            <div className="flex-1 overflow-y-auto sm:px-6">
              <div className="flex items-start justify-between sticky top-0 py-6 px-4 z-10 border-b-2 border-gray-100 bg-white">
                <h2 className="text-lg font-medium text-gray-900" id="slide-over-title">
                  Shopping cart
                </h2>
                <div className="ml-3 flex h-7 items-center">
                  <button
                    onClick={closeCart}
                    type="button"
                    className="relative -m-2 p-2 text-gray-400 hover:text-gray-500">
                    <span className="absolute -inset-0.5"></span>
                    <span className="sr-only">Close panel</span>
                    <svg
                      className="h-6 w-6"
                      fill="none"
                      viewBox="0 0 24 24"
                      strokeWidth="1.5"
                      stroke="currentColor"
                      aria-hidden="true">
                      <path strokeLinecap="round" strokeLinejoin="round" d="M6 18L18 6M6 6l12 12" />
                    </svg>
                  </button>
                </div>
              </div>

              <CartItems items={items} onRemoveItem={remove} />
            </div>

            <div className="border-t border-gray-200 px-4 py-6 sm:px-6">
              <div className="flex justify-between text-base font-medium text-gray-900">
                <p>Total</p>
                <p>{formatCurrency(getTotalPrice())}</p>
              </div>
              <div className="mt-6">
                <Link
                  to="/checkout"
                  onClick={closeCart}
                  data-testid="checkout"
                  className={`flex items-center justify-center rounded-md border border-transparent px-6 py-3 text-base font-medium shadow-sm ${
                    checkoutButtonEnabled
                      ? 'bg-indigo-600 text-white hover:bg-indigo-700'
                      : 'bg-gray-100 text-gray-800 pointer-events-none'
                  }`}>
                  Checkout
                </Link>
              </div>
              <div className="mt-6 flex justify-center text-center text-sm text-gray-500">
                <p>
                  or
                  <button
                    onClick={closeCart}
                    type="button"
                    className="font-medium text-indigo-600 hover:text-indigo-500 ml-1">
                    Continue Shopping
                    <span aria-hidden="true"> &rarr;</span>
                  </button>
                </p>
              </div>
            </div>
          </div>
        </div>
      </motion.div>
    </AnimatePresence>
  );
};
