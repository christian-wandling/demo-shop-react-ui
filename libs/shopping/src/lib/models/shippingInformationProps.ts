import { UseFormReturn } from 'react-hook-form';
import { ShippingInformationFormData } from './shippingInformationFormSchema';

export interface ShippingInformationProps {
  methods: UseFormReturn<ShippingInformationFormData>;
  onUpdateUser: () => void;
}
