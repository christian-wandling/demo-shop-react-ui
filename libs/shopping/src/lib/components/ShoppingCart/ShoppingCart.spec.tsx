import { beforeEach, describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { ShoppingCart } from './ShoppingCart';
import { useShoppingCartStore } from '../../+state/useShoppingCartStore';
import { CartItemsProps } from '../../models/cartItemsProps';

vi.mock('react-router', () => ({
  Link: ({ children, onClick }: any) => (
    <a href="#" onClick={onClick}>
      {children}
    </a>
  ),
}));

vi.mock('framer-motion', () => ({
  AnimatePresence: ({ children }: any) => <div data-testid="animate-presence">{children}</div>,
  motion: {
    div: ({ children, ...props }: any) => (
      <div data-testid="motion-div" {...props}>
        {children}
      </div>
    ),
  },
}));

vi.mock('../CartItems/CartItems', () => ({
  CartItems: ({ items, onRemoveItem }: CartItemsProps) => <div data-testid="cart-items"></div>,
}));

vi.mock('@demo-shop-react-ui/shopping', () => ({
  useShoppingCartStore: vi.fn(),
}));

vi.mock('@demo-shop-react-ui/shared', () => ({
  formatCurrency: (amount: number) => `$${amount.toFixed(2)}`,
}));

describe('ShoppingCart', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders nothing when showCart is false', () => {
    useShoppingCartStore.setState({ showCart: false });

    const { container } = render(<ShoppingCart />);
    expect(container.firstChild).toBeNull();
  });

  it('renders the shopping cart when showCart is true', () => {
    useShoppingCartStore.setState({ showCart: true });

    render(<ShoppingCart />);
    expect(screen.getByText('Shopping cart')).toBeTruthy();
  });

  it('displays the correct total price', () => {
    useShoppingCartStore.setState({ showCart: true, getTotalPrice: vi.fn().mockReturnValue(99.99) });

    render(<ShoppingCart />);
    expect(screen.getByText('$99.99')).toBeTruthy();
  });

  it('closes cart when close button is clicked', () => {
    const spy = vi.spyOn(useShoppingCartStore.getState(), 'setShowCart');
    useShoppingCartStore.setState({ showCart: true });

    render(<ShoppingCart />);
    fireEvent.click(screen.getByText('Close panel').nextElementSibling!);
    expect(spy).toHaveBeenCalledWith(false);
  });

  it('closes cart when "Continue Shopping" is clicked', () => {
    const spy = vi.spyOn(useShoppingCartStore.getState(), 'setShowCart');
    useShoppingCartStore.setState({ showCart: true });

    render(<ShoppingCart />);
    fireEvent.click(screen.getByText('Continue Shopping'));
    expect(spy).toHaveBeenCalledWith(false);
  });

  it('closes cart when Escape key is pressed', () => {
    const spy = vi.spyOn(useShoppingCartStore.getState(), 'setShowCart');
    useShoppingCartStore.setState({ showCart: true });

    render(<ShoppingCart />);
    fireEvent.keyDown(screen.getByRole('dialog'), { key: 'Escape' });
    expect(spy).toHaveBeenCalledWith(false);
  });

  describe('SnapShots', () => {
    it('should match the snapshot when cart is hidden', () => {
      useShoppingCartStore.setState({ showCart: false });

      const { baseElement } = render(<ShoppingCart />);
      expect(baseElement).toMatchSnapshot();
    });

    it('should match the snapshot when cart is in view', () => {
      useShoppingCartStore.setState({ showCart: true });

      const { baseElement } = render(<ShoppingCart />);
      expect(baseElement).toMatchSnapshot();
    });
  });
});
