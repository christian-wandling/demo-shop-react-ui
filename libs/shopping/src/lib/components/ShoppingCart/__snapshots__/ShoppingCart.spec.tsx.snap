// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShoppingCart > SnapShots > should match the snapshot when cart is hidden 1`] = `
<body>
  <div />
</body>
`;

exports[`ShoppingCart > SnapShots > should match the snapshot when cart is in view 1`] = `
<body>
  <div>
    <div
      data-testid="animate-presence"
    >
      <div
        animate="[object Object]"
        aria-labelledby="slide-over-title"
        aria-modal="true"
        class="relative z-50"
        data-testid="motion-div"
        exit="[object Object]"
        initial="[object Object]"
        role="dialog"
        transition="[object Object]"
      >
        <div
          aria-hidden="true"
          class="fixed inset-0 bg-gray-500 bg-opacity-75 transition-opacity"
        />
        <div
          class="fixed inset-0 overflow-hidden"
        >
          <div
            class="absolute inset-0 overflow-hidden"
          />
        </div>
      </div>
      <div
        animate="[object Object]"
        class="z-50 pointer-events-none fixed inset-y-0 right-0 flex max-w-full pl-10"
        data-testid="motion-div"
        exit="[object Object]"
        initial="[object Object]"
        transition="[object Object]"
      >
        <div
          class="pointer-events-auto w-screen max-w-md"
        >
          <div
            class="flex h-full flex-col bg-white shadow-xl"
          >
            <div
              class="flex-1 overflow-y-auto sm:px-6"
            >
              <div
                class="flex items-start justify-between sticky top-0 py-6 px-4 z-10 border-b-2 border-gray-100 bg-white"
              >
                <h2
                  class="text-lg font-medium text-gray-900"
                  id="slide-over-title"
                >
                  Shopping cart
                </h2>
                <div
                  class="ml-3 flex h-7 items-center"
                >
                  <button
                    class="relative -m-2 p-2 text-gray-400 hover:text-gray-500"
                    type="button"
                  >
                    <span
                      class="absolute -inset-0.5"
                    />
                    <span
                      class="sr-only"
                    >
                      Close panel
                    </span>
                    <svg
                      aria-hidden="true"
                      class="h-6 w-6"
                      fill="none"
                      stroke="currentColor"
                      stroke-width="1.5"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M6 18L18 6M6 6l12 12"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                data-testid="cart-items"
              />
            </div>
            <div
              class="border-t border-gray-200 px-4 py-6 sm:px-6"
            >
              <div
                class="flex justify-between text-base font-medium text-gray-900"
              >
                <p>
                  Total
                </p>
                <p>
                  $99.99
                </p>
              </div>
              <div
                class="mt-6"
              >
                <a
                  href="#"
                >
                  Checkout
                </a>
              </div>
              <div
                class="mt-6 flex justify-center text-center text-sm text-gray-500"
              >
                <p>
                  or
                  <button
                    class="font-medium text-indigo-600 hover:text-indigo-500 ml-1"
                    type="button"
                  >
                    Continue Shopping
                    <span
                      aria-hidden="true"
                    >
                       â†’
                    </span>
                  </button>
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
